### PR TITLE
simulation library for network added, also with selfcheck [v2]

### PR DESCRIPTION
**Pull request by clebergnu**
@jscotka this an update of your PR.  I decided to keep the changes on separate commits because the original commit is yours, and it's easier to understand what I've done.

That are still some tasks to do, some which I can remember:
 * graceful degradation: make the module *not* require `brctl`
 * graceful degradation: make it work with `sudo`

Changes from v1 (#2056):
 * style fixes
 * move tests into selftests

### CODE REVIEW

**Comment by jscotka** [comment]
Thanks a lot @clebergnu for improve this PR for merging.

**Comment by jscotka** [comment]
Hi @lmr , yes it make sense too. It depends what you prefer, tt is up to you and your decision. I have no idea what is better and how to do that. but I prefer, to have it inside avocado.utils.
because I have also second similar library for disc simulation, and this is done via iSCSI initiator/target, so that it has to have access to avocado.util.iscsitarget library
https://github.com/cockpit-project/cockpit/blob/master/test/avocado/libdisc.py mentioned in 
https://github.com/avocado-framework/avocado/issues/2008
it also allow to use loopback as disc simulation, but iSCSI is better one (closest to real disc than loopback device)

**Comment by jscotka** [comment]
Otherwise @clebergnu I have also idea to add there support for network namespaces, to completely isolate network environment, but we can start with this and later add support for network namespaces if aviable. 
Like in project https://github.com/crossdistro/network-testing/blob/master/network_testing/test_suite.py#L256
initially created for testing fedora networking stack focused on IPv6 and IPv4 fallback methods

**Comment by clebergnu** [comment]
TBH I went for the low hanging fruit during v1->v2.  I see the use cases that could benefit from this (and of course for the suggested improvements on top of this).  Still, there are a few basic things we must sort out here.

BTW, I also see `avocado.utils` as the best location (for the time being) for a module such as this.

**Comment by ldoktor** [comment]
Hello guys, I think this is similar to `lv_utils`, `disk` or `partition`, therefor yes, I believe it should go into `avocado.utils`. Anyway as @lmr I don't like the `simulation` module name. How about `network` or `networking`? That is where I'd be looking for bridge utils.

As for inclusion, I don't think we should merge this during this sprint as it adds mane new util APIs which are not stabilized yet and history showed us it will probably change several times in the first months of usage and it'd make running tests based on LTS hard to port in the future. If you really insist I'd vote for `avocado.utils.staging` namespace.

Anyway the library is really cool and I'm looking forward having it in Avocado after this LTS release.

PS: One quite important question, @jscotka did you developed it yourself, or is it based on existing code? If it's based on existing code you ought to mention it in the header (example can be found here: https://github.com/avocado-framework/avocado-misc-tests/blob/master/cpu/ebizzy.py#L17 )

**Comment by jscotka** [comment]
Hi @ldoktor. I've developed this library as part of cockpit. so not so important to mention it there.
And also I have similar library written in bash for testing components like initscripts or dnsmasq. 
and in second PR, what is based on library of my collegue I've added this info, see
https://github.com/avocado-framework/avocado/pull/2054/files#diff-89c7adf4b9e3284182ca681d1e1868e0R20

**Comment by ldoktor** [comment]
I see, anyway it is important to mention there the origin along with the author as GPL requires that so please adjust the format in the next version...

File: avocado/utils/simulation.py [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+                              ignore_status=True, verbose=False)
+            if "does not exist" not in out.stderr:
+                return True
+            else:
+                return False
+        else:
+            return True
+
+    def __createbridge(self):
+        process.run("brctl addbr %s" % self.brname)
```
Comment (ldoktor) [inline avocado/utils/simulation.py]: you should add `sudo=True` everywhere it's required (almost everywhere)

File: avocado/utils/simulation.py [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+            process.run("brctl --help", verbose=False)
+        except:
+            raise BaseException(
+                "Unable to locate brdige-utils, please install them (yum install bridge-utils)")
+
+        if self.brname and self.__checkifbridgeexist():
+            log.debug("Bridge device ADD: " + self.brname)
+            self.__createbridge()
+
+    def getName(self):
```
Comment (ldoktor) [inline avocado/utils/simulation.py]: camelCase is not the preferred style for methods in Python, it'd be really nice to adjust it to match pep style...

File: avocado/utils/simulation.py [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+
+if __name__ == '__main__':
+
+    # Test Bridge
+    br = NetworkBridge("abc")
+    br.clean()
+
+    assert "abc" not in process.run("brctl show", ignore_status=True).stdout
+    br = NetworkBridge("abc")
+    time.sleep(2)
```
Comment (ldoktor) [inline avocado/utils/simulation.py]: I'd recommend replacing these with busy-loops...

File: avocado/utils/simulation.py (hunk 2 above)
Comment (lmr) [inline avocado/utils/simulation.py]: +1000. `CamelCase` only for class names and that's about it :)

**Comment by apahim** [comment]
ping

**Comment by apahim** [comment]
Closing due to inactivity.  Please reopen if you still have interest in this feature.

**Pull request closed without merging**